### ISSUE TITLE
TS Not inferring correct type from attrs - v6

### ISSUE THREAD

**Issue by tomfridental1** [styled-components/styled-components#4076]
## Environment

`## System:
 - OS: macOS 13.3.1
 - CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
 - Memory: 186.73 MB / 16.00 GB
 - Shell: 5.9 - /bin/zsh
## Binaries:
 - Node: 16.14.0 - ~/.nvm/versions/node/v16.14.0/bin/node
 - npm: 9.2.0 - ~/.nvm/versions/node/v16.14.0/bin/npm
## npmPackages:
 - babel-plugin-styled-components: 2.1.3 => 2.1.3 
 - styled-components: 6.0.2 => 6.0.2 `

## Reproduction

v6 not inferring required props from attrs

## Steps to reproduce
```
const Test = styled.div.attrs<{ foo: number, bar: number }>({ foo: 1 })`

`

<Test bar={2} />
```

You get this error:
`No overload matches this call.
  Overload 1 of 2, '(props: PolymorphicComponentProps<"web", FastOmit<Substitute<Substitute<DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>, Omit<DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>, "ref"> & { ...; }>, { ...; }>, never>, void, void, {}, {}>): Element', gave the following error.
    Property 'foo' is missing in type '{ bar: number; }' but required in type 'FastOmit<Substitute<FastOmit<Substitute<Substitute<DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>, Omit<DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>, "ref"> & { ...; }>, { ...; }>, never>, FastOmit<...>>, keyof ExecutionProps>'.`

## Expected Behavior
Only bar is required in tsx file.

## Actual Behavior
Both foo and bar are required.

**Comment by partiellkorrekt** [comment]
I tried to fix this using [patch-package](https://github.com/ds300/patch-package). The following patch seems to work:

```
diff --git a/node_modules/styled-components/native/dist/constructors/constructWithOptions.d.ts b/node_modules/styled-components/native/dist/constructors/constructWithOptions.d.ts
index d598842..4ffac9b 100644
--- a/node_modules/styled-components/native/dist/constructors/constructWithOptions.d.ts
+++ b/node_modules/styled-components/native/dist/constructors/constructWithOptions.d.ts
@@ -9,9 +9,10 @@ type AttrsResult<T extends Attrs<any>> = T extends (...args: any) => infer P ? P
 type AttrsTarget<R extends Runtime, T extends Attrs<any>, FallbackTarget extends StyledTarget<R>, Result extends ExecutionProps = AttrsResult<T>> = Result extends {
     as: infer RuntimeTarget;
 } ? RuntimeTarget extends KnownTarget ? RuntimeTarget : FallbackTarget : FallbackTarget;
+type PartialBy<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>
 export interface Styled<R extends Runtime, Target extends StyledTarget<R>, OuterProps extends object, OuterStatics extends object = BaseObject> {
     <Props extends object = BaseObject, Statics extends object = BaseObject>(initialStyles: Styles<Substitute<OuterProps, NoInfer<Props>>>, ...interpolations: Interpolation<Substitute<OuterProps, NoInfer<Props>>>[]): IStyledComponent<R, Substitute<OuterProps, Props>> & OuterStatics & Statics;
-    attrs: <Props extends object = BaseObject, PrivateMergedProps extends object = Substitute<OuterProps, Props>, PrivateAttrsArg extends Attrs<PrivateMergedProps> = Attrs<PrivateMergedProps>, PrivateResolvedTarget extends StyledTarget<R> = AttrsTarget<R, PrivateAttrsArg, Target>>(attrs: PrivateAttrsArg) => Styled<R, PrivateResolvedTarget, PrivateResolvedTarget extends KnownTarget ? Substitute<Substitute<OuterProps, React.ComponentPropsWithRef<PrivateResolvedTarget>>, Props> : PrivateMergedProps, OuterStatics>;
+    attrs: <Props extends object = BaseObject, PrivateMergedProps extends object = Substitute<OuterProps, Props>, PrivateAttrsArg extends Attrs<PrivateMergedProps> = Attrs<PrivateMergedProps>, PrivateResolvedTarget extends StyledTarget<R> = AttrsTarget<R, PrivateAttrsArg, Target>>(attrs: PrivateAttrsArg) => Styled<R, PrivateResolvedTarget, PrivateResolvedTarget extends KnownTarget ? PartialBy<Substitute<Substitute<OuterProps, React.ComponentPropsWithRef<PrivateResolvedTarget>>, Props>, keyof PrivateAttrsArg> : PrivateMergedProps, OuterStatics>;
     withConfig: (config: StyledOptions<R, OuterProps>) => Styled<R, Target, OuterProps, OuterStatics>;
 }
 export default function constructWithOptions<R extends Runtime, Target extends StyledTarget<R>, OuterProps extends object = Target extends KnownTarget ? React.ComponentPropsWithRef<Target> : BaseObject, OuterStatics extends object = BaseObject>(componentConstructor: IStyledComponentFactory<R, StyledTarget<R>, object, any>, tag: StyledTarget<R>, options?: StyledOptions<R, OuterProps>): Styled<R, Target, OuterProps, OuterStatics>;
```

To use it, save that file to `patches/styled-components+6.0.2.patch`.

I have not checked, if that covers all the edge-cases but it seems like a good start.

**Comment by tomfridental1** [comment]
Any update on this one? 🙈

**Comment by Tokimon** [comment]
I have a similar use-case except a bit the other way around. I have a component that extends another component:

```ts
const MyExtended = styled(MyComponent).attrs({ number: 1 })`
/* styling */
`
```

Normal use of my component is like so:
```tsx
<MyComponent number={10} />
```

And with the pre-setting of the attributes/props I would expect to be able to just do this:
```tsx
<MyExtended />
```

but now TypeScript complains that the required property `number` is not set, even though I have set it with `.attrs`. I even tried this:
```ts
const MyExtended = styled(MyComponent).attrs({ number: 1 })<Omit<MyComponent, 'number'>>`
/* styling */
`
```

But the problem stays the same.

**Comment by takurinton** [comment]
Hello.

Maybe the behavior you are encountering is by design.
Actually, the notation you've shown seems to have been possible up to v5, but from v6 onwards, it has become stricter and appears to be prohibited.


Something close to this is what we are talking about here. [this comment](https://github.com/styled-components/styled-components/pull/4014/commits/e1d887b4a058e1be81da569d4cbcaf1655844c16#r1199794505)

> Allowing foo2 is a bug, we don't want to allow unknown props to be returned from attrs. It should be a partial of known props and things that React generally allows for any component.



Your expected behavior is tested here. Any properties that you don't pass to the component are represented here as optional, and your code should also do this.


https://github.com/takurinton/styled-components/blob/b358494663e66a66ea62d72487ca38baeedc4325/packages/styled-components/src/test/types.tsx#L224-L236


i.e. your code should be written like this:

```ts
// foo is optional
const Test = styled.div.attrs<{ foo?: number, bar: number }>({ foo: 1 })`
  /* your styling */
`;

<Test bar={2} />
```

**Comment by tomfridental1** [comment]
I see what you are saying but foo is not a "unknown" prop, its defined in the type and I see no reason to prevent passing required props via attrs. In the real usecase im styling already defined styled component with defined type and I cannot make foo optional as I inherit it from the original component.

More over there is a potential for bug with your suggestion, as some other developer may pass foo directly to the component overriding my attr foo version.

**Comment by tymoxx** [comment]
> Hello.
> 
> Maybe the behavior you are encountering is by design. Actually, the notation you've shown seems to have been possible up to v5, but from v6 onwards, it has become stricter and appears to be prohibited.
> 
> Something close to this is what we are talking about [this comment](https://github.com/styled-components/styled-components/pull/4014/commits/e1d887b4a058e1be81da569d4cbcaf1655844c16#r1199794505)
> 
> > Allowing foo2 is a bug, we don't want to allow unknown props to be returned from attrs. It should be a partial of known props and things that React generally allows for any component.
> 
> Your expected behavior is tested here. Any properties that you don't pass to the component are represented here as optional, and your code should also do this.
> 
> https://github.com/styled-components/styled-components/blob/b358494663e66a66ea62d72487ca38baeedc4325/packages/styled-components/src/test/types.tsx#L224-L236
> 
> i.e. your code should be written like this:
> 
> ```ts
> // foo must be optional
> const Test = styled.div.attrs<{ foo?: number, bar: number }>({ foo: 1 })`
>   /* your styling */
> `;
> 
> <Test bar={2} />
> ```

This answer makes more confusion:
1. Looking at the test example, we are more interested in `bar` property, since `bar` is required. It would be more clear they added `bar` in the **attrs**, to see how TS will reacts when it is not specified later. But such scenario is not covered.
2. There is a mistake in the comment "// Providing required prop foo, which makes it optional", since foo is optional.

**Comment by tomfridental1** [comment]
Any chance someone can look into this?

**Comment by elemanhillary** [comment]
i solved mine using

styled.div.attrs<T>((props: T) => props)`

**Comment by 4ndr345** [comment]
@tomfridental1 Looking a the test I found a workaround or maybe this is the indented way. 

https://github.com/styled-components/styled-components/blob/b358494663e66a66ea62d72487ca38baeedc4325/packages/styled-components/src/test/types.tsx#L241-L255

The "trick" is to substitute the required property with an optional one. Here is an short example. 

```typescript
const MyComponent = styled.div<{foo: number}>``;
const MyExtendedComponent = styled(MyComponent).attrs<{ foo?: number }>({ foo: 42 })``;

<MyExtendedComponent />
```

**Comment by tomfridental1** [comment]
Yeah this is a nice hack but for one thats a hack and this won't work with large interfaces with 5-10 required props

**Comment by hadnet** [comment]
TS now throws `Type 'RefObject<TextInputProps & { value: string; setNativeProps: ({ text }: { text: string; }) => void; focus: () => void; }>' is not assignable to type 'Ref<TextInput> | undefined'` when using something like this:
```tsx
export const Input = styled.TextInput.attrs(({theme}) => ({
  placeholderTextColor: theme.input.color,
}))<InputProps>`
```

**Comment by VictorGosse** [comment]
Hello,

Here is another example of what is problematic for us: https://stackblitz.com/edit/stackblitz-starters-z42ykg?file=src%2FApp.tsx
Here we can clearly see that the HTML element defined in the `attrs` is not properly inferred which prevents us from using some of the HTML attributes specific to this element such as the [download](https://www.w3schools.com/tags/att_a_download.asp) attribute for `a` element.

**Comment by elemanhillary** [comment]
> Hello,
> 
> Here is another example of what is problematic for us: https://stackblitz.com/edit/stackblitz-starters-z42ykg?file=src%2FApp.tsx Here we can clearly see that the HTML element defined in the `attrs` is not properly inferred which prevents us from using some of the HTML attributes specific to this element such as the [download](https://www.w3schools.com/tags/att_a_download.asp) attribute for `a` element.

https://stackblitz.com/edit/stackblitz-starters-guyeup?file=src%2FApp.tsx

here is a modified version that works

**Comment by VictorGosse** [comment]
@elemanhillary Yes indeed thanks but as soon as I'll change the `as` props directly when calling the component it'll be broken again `<Link as="p" download="test.csv">This is a link</Link>`

**Comment by elemanhillary** [comment]
Then you have to change the instrinsic element in types

**Comment by fpapado** [comment]
Depending on how you use `.attrs`, you can get away with a wrapper, without patching the types themselves. This requires a few acrobatics to line things up, but it is a workable alternative.

The gist of it is this:
- Making any props that overlap `.attrs` optional in the base props
- Wrapping `styled()` in a way that preserves the default semantics

```tsx
import { styled } from 'styled-components';
import type { WebTarget } from 'styled-components/dist/types';

type MakeOverlappingPropsOptional<TBase, TOther extends Partial<TBase>> = {
  [K in keyof TOther]?: K extends keyof TBase ? TBase[K] : never;
} & Omit<TBase, keyof TOther>;

export const styledWithAttrs = <TTarget extends WebTarget>(target: TTarget) => {
  // Think of this as "what the component props would have been, had
  // the styled function been called with a specific parameter
  // https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-7.html#instantiation-expressions
  type StyledForTarget = typeof styled<TTarget>;

  type ResolvedComponent =
    ReturnType<ReturnType<StyledForTarget>>;

  type Props = React.ComponentProps<ResolvedComponent>;

  return <TAttrs extends Partial<Props>>(attrs: TAttrs) => {
    // Return the regular styled() call, but overriding the props. TypeScript will
    // infer the correct expected types for the resulting component now. We have
    // to use "as any" to convince TypeScript that our attrs are OK. This might
    // preclude some advanced cases/shapes of attrs, but is good enough for
    // common use-cases and migration.
    return styled(target).attrs<MakeOverlappingPropsOptional<Props, TAttrs>>(attrs as any);
  };
};

// Usage
const StyledThing = styledWithAttrs(Thing)({something: 'value'})
```

I am not suggesting that this is the end-all-be-all solution, but it might help someone with similar constraints (no patching, object attrs usage, web target) 😌

**Comment by laboskin** [comment]
Since https://github.com/styled-components/styled-components/pull/4288 was reverted in v6.1.10, this issue is still relevant

**Comment by g-wozniak** [comment]
Any update on this ticket? Did you manage to solve it?

**Comment by tomfridental1** [comment]
Unfortunately no, the issue should be reopen @quantizor

**Comment by TMK04** [comment]
Accepts attrs(fn); no `as any`
```tsx
// Adapted from https://github.com/styled-components/styled-components/issues/4076#issuecomment-1999809785
import { ComponentProps, JSXElementConstructor } from 'react'
import { styled, StyledInstance } from 'styled-components'
import type {
  AnyComponent,
  Attrs,
  BaseObject,
  FastOmit,
  KnownTarget,
  Runtime,
  StyledTarget,
} from 'styled-components/dist/types'

type MakeOverlappingPropsOptional<
  GBase extends BaseObject,
  GOther extends Partial<GBase>
> = FastOmit<GBase, keyof GOther> & Partial<GOther>

/**
 * styled + .attrs, but with TypeScript inference
 */
function styledWithAttrs<
  GRuntime extends Runtime,
  /**
   * Provides no benefit to use string (web-components)
   */
  GTarget extends Extract<StyledTarget<GRuntime>, AnyComponent | KnownTarget>
>(target: GTarget) {
  type TBase = ComponentProps<GTarget>
  const styledTarget = styled(target)

  /**
   * @returns the regular styled() call, but overriding the props
   */
  function withAttrs<GOther extends Partial<TBase>>(attrs: Attrs<GOther>) {
    type TOverlapped = MakeOverlappingPropsOptional<TBase, GOther>
    return (
      styledTarget as StyledInstance<
        GRuntime,
        JSXElementConstructor<TOverlapped>,
        TOverlapped
      >
    ).attrs(attrs as TOverlapped)
  }

  return withAttrs
}

export default styledWithAttrs
```

**Comment by phiter** [comment]
This issue is still relevant. The type is not properly inferred. Also, I can't use `data-*` props with the `attrs` function. It's probably unrelated to this specific bug but it is still pretty annoying.

**Comment by prathamVaidya** [comment]
Issue still exists in version `6.1.19`, I suggest reopening this issue.

For now I made the attrs types optional to make it work.

Example: 

```jsx
const StyledIcon = styled(Icon).attrs<{
  name?: string;
  size?: string;
}>({
  name: 'triangle-down',
  size: 'extra-small',
})(
  () => `
    width: 12px;
  `
);
```

**Comment by nzaytsev** [comment]
I have the same issue

**Comment by eakarpov** [comment]
I have the same issue in 6.3.8.
The issue should be reopened, @quantizor

**Comment by quantizor** [comment]
## Research summary

This is blocked on TypeScript partial type parameter inference (microsoft/TypeScript#10571, PR #26349—still open since 2018). No CSS-in-JS library has solved it.

**What works today**: Declare attrs-provided props as optional in the generic:
```tsx
const Comp = styled.div.attrs<{ foo?: number; bar: string }>({ foo: 42 })``;
<Comp bar="hello" />  // foo is optional ✓
```

**What doesn't work**: Declaring `foo: number` as required and expecting attrs to auto-make it optional. This would require TS to compute which keys were satisfied by attrs at the type level—which requires partial inference that TS doesn't support.

**Other libraries**: Emotion doesn't have `.attrs()`. MUI/Chakra use `defaultProps` instead, sidestepping the issue.

Keeping open as a tracking issue for when TS ships partial inference.

**Comment by davidkarlsson** [comment]
I'm curious to why this worked with the 5.x types then if it's blocked by a TypeScript limitation?

**Comment by quantizor** [comment]
Great question. Looking at v5's `@types/styled-components`, they solved this with an explicit `A extends keyof any` type parameter that **tracks which keys were provided by attrs**:

```ts
// v5 approach: A accumulates attrs-provided key names
type MakeAttrsOptional<C, O, A extends keyof P, P = ComponentPropsWithRef<C>> =
  OmitU<P & O, A> & Partial<PickU<P & O, A>>;

// .attrs() adds keys to A
attrs<U, NewA>(attrs: Attrs<...>): ThemedStyledFunction<C, T, O & NewA, A | keyof NewA>;
//                                                                       ^^^^^^^^^^^^^^
```

This isn't partial type inference—it's a separate generic parameter that explicitly accumulates which keys become optional. v6 dropped this `A` parameter in the type system rewrite.

So this **is** fixable by reintroducing the attrs-key tracking parameter. My earlier comment about TS partial inference being the blocker was wrong—the v5 approach uses a different (working) pattern. Will investigate reintroducing it.

**Comment by quantizor** [comment]
This fix is available in `styled-components@6.4.0-prerelease.0`. Please try it and report any issues:

```
npm install styled-components@6.4.0-prerelease.0
```

**Comment by davidkarlsson** [comment]
This works great in my tests with version 6.4.0-prerelease.1!
